### ISSUE TITLE
Small Critical Error: Cannot download differentially, fallback to full download: Error: Maximum allowed size is 5 MB

### ISSUE THREAD

**Issue by orgertot** [electron-userland/electron-builder#3684]
* **Version**: 20.38.5

* **Target**: Windows

Error: Cannot download differentially, fallback to full download: Error: Maximum allowed size is 5 MB.

This Error is only thrown for a **private** Github repository. If I set it to public again this error is not thrown. Probably a small bug somewhere... I cannot figure out whats wrong :disappointed: 

```

Checking for update
Found version 1.0.6 (url: dallasspur-setup-1.0.6.exe)
Downloading update from dallasspur-setup-1.0.6.exe
updater cache dir: C:\Users\harry\AppData\Local\dallasspur-updater
No cached update info available
**Download block maps (old: "https://api.github.com/repos/harryhahn/DallasTrack/releases/assets/10233924.blockmap", new: https://api.github.com/repos/harryhahn/DallasTrack/releases/assets/10233924.blockmap)
c
at Object.responseHandler (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\src\httpExecutor.ts:207:26)
at ClientRequest.request.createRequest.response (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\src\httpExecutor.ts:267:17)
at ClientRequest.emit (events.js:182:13)
at URLRequest.ClientRequest.urlRequest.on (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\electron.asar\browser\api\net.js:207:12)
at URLRequest.emit (events.js:182:13)
From previous event:
at ElectronHttpExecutor.downloadToBuffer (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\src\httpExecutor.ts:178:60)
at C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:145:46
at Generator.next ()
From previous event:
at downloadBlockMap (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:144:29)
at C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:166:85
at Generator.next ()
From previous event:
at NsisUpdater.differentialDownloadInstaller (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:134:173)
at Object. (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:34:42)
at Generator.next ()
From previous event:
at Object.task (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:31:11)
at C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:609:25
From previous event:
at NsisUpdater.executeDownload (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:544:68)
at NsisUpdater.executeDownload (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\BaseUpdater.ts:27:18)
at NsisUpdater.doDownloadUpdate (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:27:17)
at NsisUpdater.downloadUpdate (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:419:19)
at C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:382:49
at Generator.next ()
at runCallback (timers.js:696:18)
at tryOnImmediate (timers.js:667:5)
at processImmediate (timers.js:649:5)
From previous event:
at NsisUpdater.doCheckForUpdates (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:359:34)
at NsisUpdater.checkForUpdates (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:220:35)
at App. (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\main.js:88:15)
at App.emit (events.js:187:15)
Error: Error: Cannot download "https://api.github.com/repos/harryhahn/DallasTrack/releases/assets/10233924", status 404: Not Found
at ClientRequest.request.createRequest.response (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\src\httpExecutor.ts:247:26)
at ClientRequest.emit (events.js:182:13)
at URLRequest.ClientRequest.urlRequest.on (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\electron.asar\browser\api\net.js:207:12)
at URLRequest.emit (events.js:182:13)
From previous event:
at ElectronHttpExecutor.download (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\electronHttpExecutor.ts:23:73)
at Object. (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:35:35)
From previous event:
at Object.task (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:31:11)
at C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:609:25
From previous event:
at NsisUpdater.executeDownload (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:544:68)
at NsisUpdater.executeDownload (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\BaseUpdater.ts:27:18)
at NsisUpdater.doDownloadUpdate (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:27:17)
at NsisUpdater.downloadUpdate (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:419:19)
at C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:382:49
at Generator.next ()
at runCallback (timers.js:696:18)
at tryOnImmediate (timers.js:667:5)
at processImmediate (timers.js:649:5)
From previous event:
at NsisUpdater.doCheckForUpdates (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:359:34)
at NsisUpdater.checkForUpdates (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:220:35)
at App. (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\main.js:88:15)
at App.emit (events.js:187:15)
Unhandled rejection Error: Cannot download "https://api.github.com/repos/harryhahn/DallasTrack/releases/assets/10233924", status 404: Not Found
at ClientRequest.request.createRequest.response (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\src\httpExecutor.ts:247:26)
at ClientRequest.emit (events.js:182:13)
at URLRequest.ClientRequest.urlRequest.on (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\electron.asar\browser\api\net.js:207:12)
at URLRequest.emit (events.js:182:13)
From previous event:
at ElectronHttpExecutor.download (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\electronHttpExecutor.ts:23:73)
at Object. (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:35:35)
From previous event:
at Object.task (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:31:11)
at C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:609:25
From previous event:
at NsisUpdater.executeDownload (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:544:68)
at NsisUpdater.executeDownload (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\BaseUpdater.ts:27:18)
at NsisUpdater.doDownloadUpdate (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:27:17)
at NsisUpdater.downloadUpdate (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:419:19)
at C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:382:49
at Generator.next ()
at runCallback (timers.js:696:18)
at tryOnImmediate (timers.js:667:5)
at processImmediate (timers.js:649:5)
From previous event:
at NsisUpdater.doCheckForUpdates (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:359:34)
at NsisUpdater.checkForUpdates (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:220:35)
at App. (C:\Users\harry\AppData\Local\Programs\dallasspur\resources\app.asar\main.js:88:15)
at App.emit (events.js:187:15)
```

**Comment by BungeeDesign** [comment]
I'm also getting this error. For some reason, it fails to download the difference between the current version and the latest one.  And then resorts to downloading the entire app but in most cases, it's over 5MB so it fails due to the limit.

**Comment by JonKelley88** [comment]
Same problem here with a Windows target and a private repo.

**Comment by JonKelley88** [comment]
@orgertot  @BungeeDesign 

I just got it to work by reverting `electron-builder` to version `20.38.0` and `electron-updater` to version `4.0.4`. I'm not sure which was causing the problem.

**Comment by JamiesonRoberts** [comment]
Any update on what might be causing this?

**Comment by grantcv1** [comment]
I have reverted everything in my environment back as suggested above and I'm still getting this problem. I am using a private GitHub repo with GH_TOKEN set. Windows is the target. Both my current installation and my new update are built with the same configuration as below:

electron: 5.0.0
electron-builder: 20.38.0
electron-updater: 4.0.4

    "nsis": {
      "installerIcon": "images/LegisPro.ico",
      "installerHeaderIcon": "images/LegisPro.ico",
      "createDesktopShortcut": true,
      "license": "LICENSE.txt",
      "oneClick": false,
      "allowElevation": true,
      "allowToChangeInstallationDirectory": true,
      "differentialPackage": true,
      "include": "build/installer.nsh",
      "publish": {
        "provider": "github",
        "repo": "LegisPro-Sunrise",
        "owner": "Xcential-Corporation",
        "vPrefixedTagName": true,
        "private": true,
        "releaseType": "draft"
      }
    }

  autoUpdater.setFeedURL({
    provider: "github",
    owner: "Xcential-Corporation",
    repo: "LegisPro-Sunrise",
    token: process.env.GH_TOKEN
  });
  autoUpdater.allowPrerelease = false; // Letting default errors on old releases without latest.yml files
  autoUpdater.checkForUpdatesAndNotify();

[2019-05-05 19:33:56.521] [info] # Checking for update
[2019-05-05 19:33:59.059] [info] # Found version 3.1.10-beta.10p (url: legispro-setup-3.1.10-beta.10p.exe)
[2019-05-05 19:33:59.068] [info] # Downloading update from legispro-setup-3.1.10-beta.10p.exe
[2019-05-05 19:33:59.279] [info] # No cached update info available
[2019-05-05 19:33:59.291] [info] # Download block maps (old: "https://api.github.com/repos/Xcential-Corporation/LegisPro-Sunrise/releases/assets/12401099.blockmap", new: https://api.github.com/repos/Xcential-Corporation/LegisPro-Sunrise/releases/assets/12401099.blockmap)
[2019-05-05 19:33:59.528] [error] # Error: Cannot download differentially, fallback to full download: Error: Cannot download "https://api.github.com/repos/Xcential-Corporation/LegisPro-Sunrise/releases/assets/12401099.blockmap", status 404: Not Found
    at ClientRequest.<anonymous> (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\src\httpExecutor.ts:247:26)
    at ClientRequest.emit (events.js:194:13)
    at URLRequest.<anonymous> (C:\Program Files\LegisPro\resources\electron.asar\browser\api\net.js:207:12)
    at URLRequest.emit (events.js:194:13)
From previous event:
    at ElectronHttpExecutor.downloadToBuffer (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\src\httpExecutor.ts:178:60)
    at C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:146:46
    at Generator.next (<anonymous>)
From previous event:
    at downloadBlockMap (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:145:29)
    at C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:163:51
    at Generator.next (<anonymous>)
From previous event:
    at NsisUpdater.differentialDownloadInstaller (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:135:210)
    at Object.<anonymous> (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:35:42)
    at Generator.next (<anonymous>)
From previous event:
    at Object.task (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:32:11)
    at C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:610:25
From previous event:
    at NsisUpdater.executeDownload (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:545:68)
    at NsisUpdater.executeDownload (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\BaseUpdater.ts:27:18)
    at NsisUpdater.doDownloadUpdate (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\NsisUpdater.ts:28:17)
    at NsisUpdater.downloadUpdate (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:420:19)
    at C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:383:49
    at Generator.next (<anonymous>)
    at processImmediate (internal/timers.js:443:21)
From previous event:
    at NsisUpdater.doCheckForUpdates (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:360:34)
    at NsisUpdater.checkForUpdates (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:220:35)
    at NsisUpdater.checkForUpdatesAndNotify (C:\Program Files\LegisPro\resources\app.asar\node_modules\electron-updater\src\AppUpdater.ts:249:41)
    at App.<anonymous> (C:\Program Files\LegisPro\resources\app.asar\Workspace\scripts\Main.js:79:15)
    at App.emit (events.js:199:15)

**Comment by develar** [comment]
Fixed, please see https://github.com/electron-userland/electron-builder/issues/4018

**Comment by develar** [comment]
Duplicates https://github.com/electron-userland/electron-builder/issues/3564

**Comment by Topppy** [comment]
same error. and  this error will block download progress. Is there any config that I can use to disable differentialDownloadInstaller ?

**Comment by mic1983** [comment]
Same error. Github private repository
Windows nsis build
node v10.9.0

"electron-builder": "21.2.0",
"electron-updater": "^4.1.2"
"electron": "6.0.2",

**Comment by punnadittr** [comment]
I believe it is fixed by increasing the maximum size allowed to 50MB.
But my file is 150MB so this is still an issue for me.
Why is there size limit? I'm sorry but I don't quite understand.

Now using 20.34.0 and the file is downloaded and installed properly despite errors.

**Comment by karahan** [comment]
Worked for me when i downgraded electron-updater to "3.2.3"

**Comment by almamlaka** [comment]
@karahan happy for you what is your electron-builder version ?

**Comment by almamlaka** [comment]
thank you @JonKelley88 it works 

> I just got it to work by reverting electron-builder to version 20.38.0 and electron-updater to version 4.0.4. I'm not sure which was causing the problem.

I was trying with latest version electron-builder publish 4 days ago 22.3.2 and electron-updater@4.2.0 latest without succeed 2 weeks I am trying to see which version will work!

**Comment by hmpargi** [comment]
```
`[2020-02-13 22:17:34.744] [info] Checking for update
[2020-02-13 22:17:37.465] [info] Found version 1.0.0 (url: psstracker-Setup-1.0.0.exe)
[2020-02-13 22:17:37.465] [info] Downloading update from psstracker-Setup-1.0.0.exe
[2020-02-13 22:17:37.469] [info] No cached update info available
[2020-02-13 22:17:37.471] [info] Download block maps (old: "https://api.github.com/repos/myusername/psstracker/releases/assets/18021931.blockmap", new: https://api.github.com/repos/myusername/psstracker/releases/assets/18021931.blockmap)
[2020-02-13 22:17:37.934] [error] Cannot download differentially, fallback to full download: Error: Maximum allowed size is 50 MB
    at Object.responseHandler (C:\Users\har\Desktop\Myapps\epssapp\psstracker\dist\win-unpacked\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\out\httpExecutor.js:239:26)
    at ClientRequest.<anonymous> (C:\Users\har\Desktop\Myapps\epssapp\psstracker\dist\win-unpacked\resources\app.asar\node_modules\electron-updater\node_modules\builder-util-runtime\out\httpExecutor.js:298:17)
    at ClientRequest.emit (events.js:210:5)
    at SimpleURLLoaderWrapper.<anonymous> (electron/js2c/browser_init.js:2510:12)
    at SimpleURLLoaderWrapper.emit (events.js:210:5)`
```

So why this happening i am using windows 10. my app doesn't update and throws those errors.

**Comment by Lagily** [comment]
also having this issue using 4.2.0 generic https provider

**Comment by coolswitch** [comment]
我遇到了同样的问题

```
2020-07-14 17:40:31 INFO AutoUpdate: No cached update info available
2020-07-14 17:40:31 INFO AutoUpdate: Download block maps (old: "https://...exe.blockmap", new: https://....blockmap)
2020-07-14 17:40:32 DEBUG AutoUpdate: [
  {
    "kind": 0,
    "start": 0,
    "end": 93151072
  }
]
2020-07-14 17:40:32 INFO AutoUpdate: Full: 90,967.84 KB, To download: 0 KB (0%)
2020-07-14 17:40:34 ERROR AutoUpdate: Cannot download differentially, fallback to full download: Error: sha512 checksum mismatch, expected mIVIoQ+sJcOqxBZn+QyGgtiw6i27k+LmkQo3j7BCAYlOgu9vh1yHoTMVtmqMQkWCOz8Kx+yjHUFD1+mrVfTfZw==, got Kk2y541sEKXdausV3bMTTJy/L00PhlyJ9R2gESrFZLIdwtglXXBYOCsiYM4SfHYy13vo5j49WX961KmL0er7tQ==
    at t.newError (C:\Program Files\Pitaya\resources\app.asar\dist\main.js:1:1408)
    at y.validate (C:\Program Files\Pitaya\resources\app.asar\dist\main.js:1:157342)
    at WriteStream.<anonymous> (C:\Program Files\Pitaya\resources\app.asar\dist\main.js:1:176632)
    at WriteStream.emit (events.js:223:5)
    at internal/fs/streams.js:218:14
    at C:\Program Files\Pitaya\resources\app.asar\dist\main.js:1:5818
    at FSReqCallback.oncomplete (fs.js:146:23)
```

**Comment by jo-cor** [comment]
TL;DR; Solved by adding the MIME type to IIS server.

```
- Electron v 27.0.0 [Stable on 2023-08-10], [EOL on 2024-04-16], https://www.electronjs.org/docs/latest/tutorial/electron-timelines
- Electron-Builder v 24.9.1
- Electron-Updater v 6.1.7 [latest as of 2023-Dec]
```

In my local development environment I set-up an IIS site to serve the published files, the app detects changes and attempts to download it, eventually it does it but fallbacks to a full download

```
[2023-12-14 08:32:22.187] [error] Cannot download differentially, fallback to full download: Error: Cannot download "http://localhost/app/XXXXX.exe.blockmap", status 404: Not Found
    at ClientRequest.<anonymous> (C:\Program Files\XXXXX\XXXXX\resources\app.asar\node_modules\builder-util-runtime\out\httpExecutor.js:204:34)
    at ClientRequest.emit (node:events:514:28)
    at SimpleURLLoaderWrapper.<anonymous> (node:electron/js2c/browser_init:2:49818)
    at SimpleURLLoaderWrapper.emit (node:events:514:28)
```

I added a MIME Type to the server and now it downloads it correctly in blocks

```
[2023-12-14 08:48:47.385] [info]  Download block maps (old: "http://localhost/app/XXXXX.exe.blockmap", new: http://localhost/app/XXXXX.exe.blockmap)
[2023-12-14 08:48:47.417] [info]  File has 1264 changed blocks
[2023-12-14 08:48:47.432] [info]  Full: 98,171.59 KB, To download: 25,779.55 KB (26%)
```

![image](https://github.com/electron-userland/electron-builder/assets/16821718/8a0779c3-70ef-498c-a94c-273b20542f75)

**Comment by lilian131** [comment]
@jo-cor thanks but what about using private Github repository ?
